### ISSUE TITLE
[documentation] questions on non-norwegian, key names and maps

### ISSUE THREAD

**Issue by eddy-geek** [snyball/hawck#75]
Thanks for this very pomising software :)
I'd like guidance to use hawck, and hopefully document and improve onboarding of others (which as many issues show is difficult right now :D)

In random order (as it's all related)

### Key names

**What is the list of string/integer keys can we can put in a  `key "<foo>" => write("bar")` line?**

I surmise that at low level hawck intercepts kernel event code, aka keycodes, aka [/usr/include/linux/input-event-codes.h](https://github.com/torvalds/linux/blob/master/include/uapi/linux/input-event-codes.h), that are also repeated in */usr/share/hawck/keymaps/default_linux.lua* and can be seen with `showkey`.

But it does not seem that I can use these key names or number in my scripts. for example using strings like "kp0" or "calc" or their corresponding keycodes eg `key(140)` for `calc` did not work. looking at `kbd:getKeysym` in kbd.lua at least the key should work?

=> guess:
* only key names from keymaps can be used
* tere is a bug with specifying keycodes directly?

### Default key maps

~/.local/share/hawck/cfg.lua has `keymap = "no",`  ie norwegian, by default, which should probably be changed during install process, maybe based on system configuration. At least warn the user more prominently?

### Built-in Key maps  

I'm confused overall on the role of keymaps. I guess keymaps are "syntactic sugar" to be able to refer to a trigger key by its localized label (eg in norwegian "-" will mean KEY_SLASH) ?

Also it seems fremebuffer console keymaps are used, it it just because they are a convenient built-in way to find this mapping (more convenient than eg xkb)?

for example with `keymap ="no"` then hawck goes and use /usr/share/hawck/keymaps/qwerty/no.map ... so will it map any given key string eg `key "exclam"` to keycode eg `2` (a cursory look at the Keymap.lua seems to confirm that)

if I put `keymap ="us"`  it goes to parse /usr/share/keymaps/i386/qwerty/us.kmap.gz ?

~what this `no.map` is for it looks a bit like what libinput/XKB does in UI (/usr/share/X11/xkb/symbols/no), so, is it only for console? it is unclear if anything else is supported. you mention using console-data keymaps (*.map.gz?) in some threads, but I cannot find any on my ubuntu 22.04~ => *edit: found it*

### User Key maps  

I have a custom layout that I defined myself in */usr/share/X11/xkb/symbols*. do I absolutely need to make Hawck aware of it, or can I set `keymap = "us"` and give the name of the "us" key at the same physical position for the trigger and it will work? (guess: yes I can)

**Comment by eddy-geek** [comment]
I have found */usr/share/keymaps/i386/qwerty/us.kmap.gz* which I think (looking at Keymap.lua) is what would be parsed with `keymap ="us" ` ?

**Comment by eddy-geek** [comment]
an unrelated confusing point, changing the script will autoreload, but new keys wil be silently ignored until `hawck-add` is rerun

**Comment by snyball** [comment]
> an unrelated confusing point, changing the script will autoreload, but new keys wil be silently ignored until `hawck-add` is rerun

Hawck could do a better job of warning about this using notifications. It has to do with the whitelisting, Hawck has kind of an odd security model to be able to support both arbitrary command execution (potentially evil keylogging) and to support arbitrary transformation of key inputs to outputs (safe if sandboxed.)

I have some ideas about how to replace the old Hawck security model completely with a sandbox, but I haven't had time to work on that for a while. I think it makes some wrong tradeoffs in the end.

> Default key maps

I would set this automatically, but had problems doing it portably. The easiest way would be to read `/etc/vconsole.conf`, but that file might not be present, or might not have the same keymap as in Wayland or X11 (some users never use the built-in Linux framebuffer console.)

But changing the installation process to read `/etc/vconsole.conf` might be the most portable, and additionally it might be good for inputd to warn about detecting a mismatch on startup.

> Built-in Key maps

Hawck simply uses console keymaps, but with the addition of aliases between unicode characters and key names, so that `insert "/"` is able to insert a slash. Having to type whatever name the console maps use for the keys would be inconvenient.

It doesn't do any kind of localization of key names like with your `KEY_SLASH` example. 

> XKB

I meant to transition Hawck into using the XKB definitions instead of the console ones ages ago, because of the problems you've referred to, but I never got around to doing it. It's number one on my priority list for what large changes that should be made to Hawck before any 1.0 release is ever made.

There's also a big usability issue with installing the console keymaps on Ubuntu, because it asks the user a weird question about touching keymaps or not before installing.

### In summation

All in all Hawck has a lot of glaring issues that I've become aware of over the years (some not mentioned here,) but it has been in maintenance mode for some time. I've been diligent about fixing build/deploy issues on Arch Linux (main target platform) but haven't had the bandwidth for testing lots of other distros/configurations and adding new big things™

I see you've forked the repo, feel free to contact me with any questions about the code :) (It has some vestigial code paths from when Hawck had a GUI, and a bit of NIH syndrome because C++ dependency management is hard.)

**Comment by eddy-geek** [comment]
Thanks you for taking the time to explain, it helps a lot.

Please have a look at my 4 PRs, at least the 2 bug fixes #76 and #78 :-)

One more noob question, for some reason I cannot run lua myself (although hawck and hawck-add can without issue). I get:

```sh
> cd ~/.local/share/hawck/scripts/
> lua -l init
./LLib/Hawck.lua:215: Undefined variable: _PROMPT
```

Workarounds are either to add ` -e "_PROMPT=' lua> '"` to the command, or to comment the [last line in Hawck.lua](https://github.com/snyball/Hawck/blob/master/src/Lua/Hawck.lua#L236) 

```lua
setmetatable(_G, ProtectedMeta)
```

then the error disappears and everything runs normally. Am I missing something obvious?

**Comment by snyball** [comment]
@eddy-geek I reviewed your pull requests, but you might not be getting notifications for them.
